src/Transit/Internal/Pipeline.hs

```diff
@@ -0,0 +1,132 @@
+module Transit.Internal.Pipeline
+  ( sendPipeline
+  , receivePipeline
+  -- * for tests
+  , assembleRecordC
+  )
+where
+
+import Protolude
+
+import Crypto.Hash (SHA256(..))
+import Data.Conduit ((.|))
+import Data.ByteString.Builder(toLazyByteString, word32BE)
+import Data.Binary.Get (getWord32be, runGet)
+import Crypto.Saltine.Internal.ByteSizes (boxNonce)
+
+import qualified Crypto.Hash as Hash
+import qualified Conduit as C
+import qualified Data.Conduit.Network as CN
+import qualified Data.Conduit.Binary as CB
+import qualified Data.Binary.Builder as BB
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Lazy as BL
+import qualified Crypto.Saltine.Core.SecretBox as SecretBox
+import qualified Crypto.Saltine.Class as Saltine
+
+import Transit.Internal.Network (TCPEndpoint(..), CommunicationError(..))
+import Transit.Internal.Crypto (encrypt)
+
+-- | Given the peer network socket and the file path to be sent, this Conduit
+-- pipeline reads the file, encrypts and send it over the network. A sha256
+-- sum is calculated on the input before encryption to compare with the
+-- receiver's decrypted copy.
+sendPipeline :: C.MonadResource m =>
+                FilePath
+             -> TCPEndpoint
+             -> SecretBox.Key
+             -> C.ConduitM a c m (Text, ())
+sendPipeline fp (TCPEndpoint s) key =
+  C.sourceFile fp .| sha256PassThroughC `C.fuseBoth` (encryptC key .| CN.sinkSocket s)
+
+-- | Receive the encrypted bytestream from a network socket, decrypt it and
+-- write it into a file, also calculating the sha256 sum of the decrypted
+-- output along the way.
+receivePipeline :: C.MonadResource m =>
+                   FilePath
+                -> Int
+                -> TCPEndpoint
+                -> SecretBox.Key
+                -> C.ConduitM a c m (Text, ())
+receivePipeline fp len (TCPEndpoint s) key =
+    CN.sourceSocket s
+    .| assembleRecordC
+    .| decryptC key
+    .| CB.isolate len
+    .| sha256PassThroughC `C.fuseBoth` C.sinkFileCautious fp
```

Comment:
> would be to remove the TCPEndpoint parameter and have this function construct only the tail of the pipeline. This would allow (and require) the caller to compose the pipeline with any Conduit source

Yes, totally agree. I had been thinking about this as well. I was thinking if the pipelines should just get rid of the head and tail and just have the inner computation alone. This would allow it to be independently tested with other sources and sinks.

I was disappointed not to find some sort of a "string" source and sink that would allow constructing a source from a given source. For eg: like  `stringSource "foobar"` that would give us a file descriptor, with the file contents being "foobar". Perhaps we could try proposing it in the Conduit issue tracker.